zbuf/scanner.go

```diff
@@ -163,3 +187,42 @@ func (n *namedScanner) Pull(done bool) (Batch, error) {
 	}
 	return b, err
 }
+
+type MultiScanner struct {
+	scanners []Scanner
+	progress Progress
+	current  Scanner
+}
+
+var _ Scanner = (*MultiScanner)(nil)
+
+func NewMultiScanner(scanners ...Scanner) *MultiScanner {
+	return &MultiScanner{
+		scanners: scanners,
+	}
+}
+
+func (m *MultiScanner) Pull(done bool) (Batch, error) {
+	for {
+		if m.current == nil {
+			if len(m.scanners) == 0 {
+				return nil, nil
+			}
+			m.current = m.scanners[0]
+			m.scanners = m.scanners[1:]
+		}
+		batch, err := m.current.Pull(done)
+		if err != nil {
+			return nil, err
+		}
+		if batch != nil {
+			return batch, nil
+		}
+		m.progress.Add(m.current.Progress())
+		m.current = nil
+	}
```

Comment:
Nit: Don't really need m.current.
```suggestion
	for len(m.scanners) > 0 {
		if batch, err := m.scanners[0].Pull(done); batch != nil || err != nil {
			return batch, err
		}
		m.progress.Add(m.scanners[0].Progress())
		m.scanners = m.scanners[1:]
	}
	return nil, nil
```

Reply:
You need to do repeated pulls from current until EOS before you move on to the next scanner.

Reply:
This'll do that! Clearer this way?
```go
	for len(m.scanners) > 0 {
		batch, err := m.scanners[0].Pull(done)
		if batch != nil || err != nil {
			return batch, err
		}
		m.progress.Add(m.scanners[0].Progress())
		m.scanners = m.scanners[1:]
	}
	return nil, nil
```